ingestion-beam/src/main/java/com/mozilla/telemetry/heka/HekaTransform.java

```diff
@@ -0,0 +1,33 @@
+package com.mozilla.telemetry.heka;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.io.FileIO.ReadableFile;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.json.JSONObject;
+import com.mozilla.telemetry.transforms.WithErrors;
+import com.mozilla.telemetry.transforms.MapElementsWithErrors;
+
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+
+public class HekaTransform extends MapElementsWithErrors<ReadableFile, List<PubsubMessage>> {
+    @Override
+    protected List<PubsubMessage> processElement(ReadableFile readableFile) throws IOException {
```

Comment:
We will also need to define a `processError` override method, since `MapElementsWithErrors` doesn't have a default implementation of how to process a `ReadableFile` into an error ouput pubsubmessage. We might want to turn this into a PusubMessage with no payload, but make sure we add a field for the file path of the ReadableFile, so we can go back and look at that file to see what's problematic.

Reply:
can you go into more detail with what you mean by a "field for the file path"?

Reply:
See `FailureMessage.java`. That has various methods for taking a caught exception and the associated record, and creating a PubsubMessage out of it. In this case, we'd want something like:

```
  public static PubsubMessage of(Object caller, ReadableFile readableFile, Throwable e) {
    Map<String, String> attributes = errorAttributes(caller, e);
    attributes.put("readable_file", readableFile.toString());
    return new PubsubMessage(null, errorAttributes(caller, e));
  }
```

I expect `readableFile.toString()` will give a full description of the object, including the path in GCS.

Reply:
Looks like it's failing on the `Heka.Header.parseFrom(headerBuffer)` call. `headerBuffer` is a zero-length byte array, and this raises `com.google.protobuf.InvalidProtocolBufferException: Message missing required fields: message_length`.

So that gets caught by the MapElementsWithErrors machinery, invoking `processError` and it looks like I steered you wrong there. You call `FailureMessage.of` which ends up creating a PubsubMessage with a null payload. I had forgotten that the null payload would cause encoding errors. So perhaps it would be best to make the payload `"{}".getBytes(StandardCharsets.UTF_8)`.

Reply:
Yup, this was the problem, plus a small mistake in your suggested error message code (not actually using the `attributes` map).

Reply:
Here's what an error payload looks like now: 

```
{"attributeMap":{"error_message":"com.google.protobuf.InvalidProtocolBufferException: Message missing required fields: message_length","exception_class":"com.google.protobuf.InvalidProtocolBufferException","stack_trace":"[com.google.protobuf.UninitializedMessageException.asInvalidProtocolBufferException(UninitializedMessageException.java:79), com.google.protobuf.AbstractParser.checkMessageInitialized(AbstractParser.java:68), com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:191), com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:203), com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:208), com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:48), com.mozilla.telemetry.heka.Heka$Header.parseFrom(Heka.java:575), com.mozilla.telemetry.heka.HekaReader.readHekaMessage(HekaReader.java:48), com.mozilla.telemetry.heka.HekaReader.readHekaStream(HekaReader.java:106), com.mozilla.telemetry.heka.HekaTransform.processElement(HekaTransform.java:23), com.mozilla.telemetry.heka.HekaTransform.processElement(HekaTransform.java:1), com.mozilla.telemetry.transforms.MapElementsWithErrors$DoFnWithErrors.processElementOrError(MapElementsWithErrors.java:89), com.mozilla.telemetry.transforms...","readable_file":"ReadableFile{metadata=Metadata{resourceId=/var/maven/project/ingestion-beam/tmp/cheezburger.bad.heka, sizeBytes=4, isReadSeekEfficient=true, lastModifiedMillis=0}, compression=UNCOMPRESSED}","error_type":"HekaTransform"},"payload":"e30="}
```